rclcpp_lifecycle/src/lifecycle_node_interface_impl.hpp

```diff
@@ -335,6 +365,7 @@ class LifecycleNode::LifecycleNodeInterfaceImpl
   const State &
   get_current_state()
   {
+    std::lock_guard<std::recursive_mutex> lock(state_machine_mutex_);
     current_state_ = State(state_machine_.current_state);
     return current_state_;
```

Comment:
As pointed out in the initial issue, we could also make this even safer by not storing the current_state and just using the lock.  That is, change this code to:

```
    std::lock_guard<std::recursive_mutex> lock(state_machine_mutex_);
    return State(state_machine_.current_state);
```

(and remove the `current_state_` member variable)

Reply:
> i think it should return the pointer to current state, unless it does not really indicate the current state? cz it might be changed right after that... thinking i would like to keep the current implementation.

The issue is that the returned reference can be invalidated anytime, and an external caller doesn't have acces to the `state_machine_mutex_`.
I agree with @clalancette, returning a copy will avoid the race condition altogether.

Reply:
I mentioned before, i was hesitating to do this since this affects user interfaces for `LifecycleNode`, most of operation methods return `State &`. returning constructed value as a copy can be done, but just checking if we can actually do this now. or any good solution or idea to keep the user interface?

https://github.com/ros2/rclcpp/blob/152dbc6e38e18c18a5883b51c33162d237f891e0/rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_node.hpp#L789-L909

Reply:
I think you can change those methods to return `State` by value as well.
It's probably not going to break any external code.

You'll probably need though to add a copy constructor to `State`.
The `State` class internally stores a rcl handle https://github.com/ros2/rclcpp/blob/1688f05aead3968bcea6cab27b346260f633f0b0/rclcpp_lifecycle/include/rclcpp_lifecycle/state.hpp#L94
so maybe that needs mutual exclusion internally as well (?)
The fact that we're using shallow copies might be a bit of a problem.